app/main.py

```diff
@@ -1,2 +1,167 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from fractions import Fraction
+from typing import Any, Hashable, Optional, Iterator, Iterable
+
+
 class Dictionary:
-    pass
+    _INITIAL_CAPACITY = 8
+    _LOAD_FACTOR = Fraction(2, 3)
+    _CAPACITY_MULTIPLIER = 2
+
+    _sentinel = object()
+
+    @dataclass
+    class Node:
+        key: Hashable
+        key_hash: int
+        value: Any
+
+    def __init__(self, ) -> None:
+        self._capacity = self._INITIAL_CAPACITY
+        self._threshold = int(self._capacity * self._LOAD_FACTOR)
+        self._size = 0
+        self._hash_table: list[Dictionary.Node | None] = (
+            [None] * self._capacity
+        )
+
+    def _calculate_index(self, key: Hashable) -> int:
+        index = hash(key) % self._capacity
+
+        while (
+                self._hash_table[index] is not None
+                and self._hash_table[index].key != key
+        ):
+            index = (index + 1) % self._capacity
+
+        return index
+
+    def __getitem__(self, key: Hashable) -> Any:
+        item: Dictionary.Node | None = (
+            self._hash_table[self._calculate_index(key)]
+        )
+
+        if not item:
+            raise KeyError(f"'{key}'")
+
+        return item.value
+
+    def __setitem__(self, key: Hashable, value: Any) -> None:
+        index = self._calculate_index(key)
+
+        if not self._hash_table[index]:
+            if self._size + 1 >= self._threshold:
+                self._resize_hash_table()
+                index = self._calculate_index(key)
+            self._hash_table[index] = Dictionary.Node(key, hash(key), value)
+            self._size += 1
+        else:
+            self._hash_table[index].value = value
+
+    def __delitem__(self, key: Hashable) -> None:
+        index = self._calculate_index(key)
+
+        if not self._hash_table[index]:
+            raise KeyError(f"'{key}'")
+
+        self._hash_table[index] = None
+        self._size -= 1
+
+        next_index = (index + 1) % self._capacity
+        while self._hash_table[next_index] is not None:
+            item = self._hash_table[next_index]
```

Comment:
making nested while loops is not very optimal here - you could reuse part of your functionality in  _resize_hash_table to make it work. Basically if you parametrize multiplier in resize method and call it with multiplier of 1 from here it will reset all the indexes correctly